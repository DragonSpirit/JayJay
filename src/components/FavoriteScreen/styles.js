import { StyleSheet } from 'react-native'

const styles = StyleSheet.create({
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  feedText: {
    marginBottom: 10,
  },
  readPostBtn: {
    borderRadius: 0,
    marginLeft: 0,
    marginRight: 0,
    marginBottom: 0,
  },
  scrollToTopView: {
    width: 50,
    height: 50,
    position: 'absolute',
    bottom: 10,
    right: 0,
  },
})

export default styles
