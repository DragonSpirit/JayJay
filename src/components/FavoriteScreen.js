import React from 'react'
import { View, Text } from 'react-native'
import { FavoriteTabIcon } from './TabIcons/FavoriteTabIcon'
import PropTypes from 'prop-types'
import { commonStyles } from './common.styles'

class FavoriteScreen extends React.PureComponent {
  constructor(props) {
    super(props)
  }

  render() {
    return (
      <View style={commonStyles.container}>
        <Text>
            Favorite posts
        </Text>
      </View>
    )
  }

  static navigationOptions = {
    title: 'Favorites',
    tabBarIcon: FavoriteTabIcon,
  }
  
  static propTypes = {
    posts: PropTypes.array,
  }

}
export default FavoriteScreen
