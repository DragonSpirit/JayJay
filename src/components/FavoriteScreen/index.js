// @flow

import React from 'react'
import { Text, View, FlatList, TouchableWithoutFeedback, Dimensions } from 'react-native'
import { array, bool, object } from 'prop-types'
import { FavoriteTabIcon } from '../TabIcons/FavoriteTabIcon'
import { Button, Card, Avatar } from 'react-native-elements'
import HTML from 'react-native-render-html'
import { parseTitle } from '../../helpers/StringUtil'
import styles from './styles'
import { commonStyles } from '../common.styles'
import type { Post } from '../../reducers/ReducerTypes'

type State = {
  isUpButtonShowed: boolean,
  date: Date
}

type Props = {
  favoritePosts: Array<Post>,
  navigation: Object,
  isPostsLoading?: boolean,
}

class FavoriteScreen extends React.Component<Props, State> {

  static navigationOptions = {
    title: 'Избранное',
    tabBarIcon: FavoriteTabIcon,
  }

  static propTypes = {
    favoritePosts: array,
    navigation: object,
    isPostsLoading: bool,
  };

  static defaultProps = {
    favoritePosts: [],
  }

  flatListRef: * = null;
  scrollToTop: boolean = false;
  listeners = [];

  constructor(props: Props) {
    super(props)
    this.state = {
      isUpButtonShowed: false,
      date: new Date(), // dirty hack, react not rerender after update store (redux or react-navigation issue)
    }
  }

  componentDidMount(): void {
    if (this.props.navigation) {
      this.listeners = [
        this.props.navigation.addListener('didFocus', () => {
          this.setState({
            date: new Date(),
          })
        }),
      ]
    }
  }

  componentWillUnmount(): void {
    this.listeners.forEach( item => item.remove())
  }

  reloadPostsView = () => (
    <View style={commonStyles.container}>
      <Text style={styles.welcome}>
        Список избранного пуст
      </Text>
      <Button title='Перейти к постам'
        onPress={() => this.props.navigation.navigate('Feed')}
        buttonStyle={[commonStyles.buttonStyleCommon, commonStyles.buttonStyleBig]} />
    </View>
  )

  _renderItem = ({item}) => {
    return (
      <TouchableWithoutFeedback onPress={() => this.props.navigation.navigate('Details', item)}>
        <Card
          title={`${item.author} \n ${item.title}`}
          image={item.img ? {uri: item.img} : null}>
          <HTML html={`${parseTitle(item.text)}...`} style={styles.feedText} />
        </Card>
      </TouchableWithoutFeedback>
    )
  }

  _keyExtractor = (item: Post) => `${item.did}`

  _onScroll = event => {
    if (this.scrollToTop) {
      return
    }
    if (event.nativeEvent.contentOffset.y > Dimensions.get('window').height && !this.state.isUpButtonShowed) {
      this.setState({isUpButtonShowed: true})
    }
  }

  _scrollToTop = () => {
    this.scrollToTop = true
    // $FlowFixMe
    this.flatListRef.scrollToIndex({animated: true, index: 0})
    this.setState({isUpButtonShowed: false})
    setTimeout(() => {this.scrollToTop = false}, 200)
  }

  render() {
    const {
      favoritePosts,
    } = this.props
    if (favoritePosts.length === 0)
      return this.reloadPostsView()
    return (
      <View style={commonStyles.flex}>
        <View style={commonStyles.container}>
          <FlatList
            ref={ref => {this.flatListRef = ref}}
            data={favoritePosts}
            renderItem={this._renderItem}
            onScroll={this._onScroll}
            scrollEventThrottle={64}
            keyExtractor={this._keyExtractor}
            extraData={this.state.date}
          />
        </View>
        {this.state.isUpButtonShowed ?
          <View style={styles.scrollToTopView}>
            <Avatar onPress={this._scrollToTop}
              rounded
              icon={{name: 'arrow-drop-up'}}
              width={40}
              height={40}
            />
          </View> : null
        }
      </View>)
  }
}

export default FavoriteScreen
