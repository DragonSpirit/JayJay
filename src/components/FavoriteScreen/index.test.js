import React from 'react';
import FavoriteScreen from './';

import renderer from 'react-test-renderer';

test('renders correctly', () => {
  const tree = renderer.create(<FavoriteScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly with empty feed but exists authors', () => {
  const tree = renderer.create(<FavoriteScreen authors={['test']} feed={[]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly with non empty feed and authors', () => {
  const feed = [{
    author: 'test',
    id: 1,
    did: 1,
    title: 'Test',
    text: 'Test',
    ts: 1,
    url: 'local.url',
    tags: 'tag',
    isFavorite: false,
  }]
  const tree = renderer.create(<FavoriteScreen authors={['test']} feed={feed} />).toJSON();
  expect(tree).toMatchSnapshot();
});