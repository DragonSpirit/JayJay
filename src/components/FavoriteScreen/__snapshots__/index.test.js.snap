// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "margin": 10,
        "textAlign": "center",
      }
    }
  >
    Список избранного пуст
  </Text>
  <View
    style={
      Array [
        Object {
          "marginLeft": 15,
          "marginRight": 15,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        pointerEvents="box-only"
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#9E9E9E",
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 19,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            Object {
              "padding": 12,
            },
            undefined,
            undefined,
            undefined,
            Array [
              Object {
                "backgroundColor": "rgba(92, 99,216, 1)",
                "borderColor": "transparent",
                "borderRadius": 5,
                "borderWidth": 0,
              },
              Object {
                "height": 45,
                "width": 300,
              },
            ],
            undefined,
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Array [
                Object {
                  "color": "white",
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontSize": 17.5,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
              ],
            ]
          }
        >
          Перейти к постам
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly with empty feed but exists authors 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "margin": 10,
        "textAlign": "center",
      }
    }
  >
    Список избранного пуст
  </Text>
  <View
    style={
      Array [
        Object {
          "marginLeft": 15,
          "marginRight": 15,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        pointerEvents="box-only"
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#9E9E9E",
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 19,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            Object {
              "padding": 12,
            },
            undefined,
            undefined,
            undefined,
            Array [
              Object {
                "backgroundColor": "rgba(92, 99,216, 1)",
                "borderColor": "transparent",
                "borderRadius": 5,
                "borderWidth": 0,
              },
              Object {
                "height": 45,
                "width": 300,
              },
            ],
            undefined,
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Array [
                Object {
                  "color": "white",
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontSize": 17.5,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
              ],
            ]
          }
        >
          Перейти к постам
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly with non empty feed and authors 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "margin": 10,
        "textAlign": "center",
      }
    }
  >
    Список избранного пуст
  </Text>
  <View
    style={
      Array [
        Object {
          "marginLeft": 15,
          "marginRight": 15,
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        pointerEvents="box-only"
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#9E9E9E",
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 19,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            Object {
              "padding": 12,
            },
            undefined,
            undefined,
            undefined,
            Array [
              Object {
                "backgroundColor": "rgba(92, 99,216, 1)",
                "borderColor": "transparent",
                "borderRadius": 5,
                "borderWidth": 0,
              },
              Object {
                "height": 45,
                "width": 300,
              },
            ],
            undefined,
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {},
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Array [
                Object {
                  "color": "white",
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontSize": 17.5,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
              ],
            ]
          }
        >
          Перейти к постам
        </Text>
      </View>
    </View>
  </View>
</View>
`;
